Source/Core/VideoCommon/VertexLoaderX64.cpp

```diff
@@ -0,0 +1,465 @@
+		_mm_set_ps1(1./(1u<< 4)), _mm_set_ps1(1./(1u<< 5)), _mm_set_ps1(1./(1u<< 6)), _mm_set_ps1(1./(1u<< 7)),
+		_mm_set_ps1(1./(1u<< 8)), _mm_set_ps1(1./(1u<< 9)), _mm_set_ps1(1./(1u<<10)), _mm_set_ps1(1./(1u<<11)),
+		_mm_set_ps1(1./(1u<<12)), _mm_set_ps1(1./(1u<<13)), _mm_set_ps1(1./(1u<<14)), _mm_set_ps1(1./(1u<<15)),
+		_mm_set_ps1(1./(1u<<16)), _mm_set_ps1(1./(1u<<17)), _mm_set_ps1(1./(1u<<18)), _mm_set_ps1(1./(1u<<19)),
+		_mm_set_ps1(1./(1u<<20)), _mm_set_ps1(1./(1u<<21)), _mm_set_ps1(1./(1u<<22)), _mm_set_ps1(1./(1u<<23)),
+		_mm_set_ps1(1./(1u<<24)), _mm_set_ps1(1./(1u<<25)), _mm_set_ps1(1./(1u<<26)), _mm_set_ps1(1./(1u<<27)),
+		_mm_set_ps1(1./(1u<<28)), _mm_set_ps1(1./(1u<<29)), _mm_set_ps1(1./(1u<<30)), _mm_set_ps1(1./(1u<<31)),
+	};
+
+	X64Reg coords = XMM0;
+
+	int elem_size = 1 << (format / 2);
+	int load_bytes  = elem_size * count_in;
+	int store_bytes = sizeof(float) * count_out;
+	if (load_bytes > 8)
+		MOVDQU(coords, data);
+	else if (load_bytes > 4)
+		MOVQ_xmm(coords, data);
+	else
+		MOVD_xmm(coords, data);
+	UpdateSourceOffset(attribute, load_bytes);
+
+	PSHUFB(coords, M(&shuffle_lut[format][count_in - 1]));
+
+	if (format != FORMAT_FLOAT)
+	{
+		// Sign extend
+		if (format == FORMAT_BYTE)
+			PSRAD(coords, 24);
+		if (format == FORMAT_SHORT)
+			PSRAD(coords, 16);
+
+		CVTDQ2PS(coords, R(coords));
+
+		if (scaling_exponent)
+			MULPS(coords, M(&scale_factors[scaling_exponent]));
+	}
+
+	OpArg dest = MDisp(dst_reg, m_dst_ofs);
+	if (store_bytes > 8)
+		MOVUPS(dest, coords);
+	else
+		MOVLPS(dest, coords);
+
+	native_format->components = count_out;
+	native_format->enable = true;
+	native_format->offset = m_dst_ofs;
+	native_format->type = VAR_FLOAT;
+	native_format->integer = false;
+	m_dst_ofs += store_bytes;
+	return load_bytes;
+}
+
+// TODO: rewrite fallbacks without swapping
+
+void VertexLoaderX64::ReadColor(OpArg data, u64 attribute, int format, int elements)
+{
+	int load_bytes = 0;
+	static const u32 mask_565 = 0xF8FCF800;
+	static const u32 mask_0f  = 0x0F0F0F0F;
+	static const u32 mask_f0  = 0xF0F0F0F0;
+	static const u32 mask_fc  = 0xFCFCFCFC;
+	switch (format)
+	{
+		case FORMAT_24B_888:
+		case FORMAT_32B_888x:
+		case FORMAT_32B_8888:
+			MOV(32, R(scratch1), data);
+			if (format != FORMAT_32B_8888 || !elements)
+				OR(32, R(scratch1), Imm32(0xFF000000));
+			MOV(32, MDisp(dst_reg, m_dst_ofs), R(scratch1));
+			load_bytes = 3 + (format != FORMAT_24B_888);
+			break;
+
+		case FORMAT_16B_565:
+			//                   RRRRRGGG GGGBBBBB
+			// AAAAAAAA BBBBBBBB GGGGGGGG RRRRRRRR
+			LoadAndSwap(16, scratch1, data);
+			if (cpu_info.bBMI2)
+			{
+				PDEP(32, scratch1, scratch1, M(&mask_565));
+				MOV(32, R(scratch2), R(scratch1));
+			}
+			else
+			{
+				MOV(32, R(scratch3), R(scratch1));
+				SHL(32, R(scratch1), Imm8(16));
+				AND(32, R(scratch1), Imm32(0xF8000000));
+				MOV(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
```

Comment:
(Repeat; I accidentally commented on an outdated commit.)  One of these MOVs is redundant.